src/haywire/http_server.c

```diff
@@ -268,6 +288,80 @@ void http_stream_on_read_http_parser(uv_stream_t* tcp, ssize_t nread, const uv_b
+{
+    http_connection* connection = (http_connection*)tcp->data;
+
+    // TODO: nread == 0 should mean the socket closed. We should handle this properly.
+    if (nread >= 0)
+    {
+        //printf("BEFORE\tREAD: %d BUFF_LEN: %d PREV: %d\n", nread, connection->request_buffer_length, connection->prevbuflen);
+        
+        memcpy(&connection->request_buffer[connection->request_buffer_length], buf->base, nread);
+        connection->request_buffer_length += nread;
+        int counter = 0;
+
+        //printf("AFTER\tREAD: %d BUFF_LEN: %d PREV: %d ", nread, connection->request_buffer_length, connection->prevbuflen);
+        
+        while (connection->prevbuflen < connection->request_buffer_length)
+        {
+            counter++;
+            
+            int parsed;
+            char *method;
+            char *path;
+            int minor_version;
+            struct phr_header headers[100];
+            size_t method_len;
+            size_t path_len;
+            size_t num_headers;
+            num_headers = sizeof(headers) / sizeof(headers[0]);
+            
+            parsed = phr_parse_request(&connection->request_buffer[connection->prevbuflen], connection->request_buffer_length - connection->prevbuflen,
+                                       &method, &method_len, &path, &path_len, &minor_version, headers,
+                                       &num_headers, 0);
+
+            //printf("POS: %d PARSED: %d ----------------------------------------\n", connection->prevbuflen, parsed);
+            
+            if (parsed > 0)
+            {
+                // Successfully parsed the request.
+                connection->prevbuflen += parsed;
+                            
+                connection->keep_alive = 1;
+                http_request* request = create_http_request(connection);
+                connection->request = request;
+                
+                path[path_len] = 0x00;
+                request->url = path;
+                
+                http_request_complete_request(connection);
+            }
+            else if (parsed == -1)
+            {
+                memmove(connection->request_buffer, &connection->request_buffer[connection->prevbuflen], connection->request_buffer_length - connection->prevbuflen);
+                connection->request_buffer_length = connection->request_buffer_length - connection->prevbuflen;
+                connection->prevbuflen = 0;
+                
+                //printf("BUFF_LEN: %d PREV: %d\n", connection->request_buffer_length, connection->prevbuflen);
+                break;
+            }
+            else if (parsed == -2)
+            {
+                break;
+            }
+        }
+        
+        if (connection->prevbuflen == connection->request_buffer_length)
+        {
+            connection->request_buffer_length = 0;
+            connection->prevbuflen = 0;
+        }
+    }
```

Comment:
Not closing the connection here causes the `Assertion `((stream)->io_watcher.fd) >= 0' failed.` error, as the connections are kept on CLOSE_WAIT after the client closes on their end.

Example:

```
[ec2-user@ip-172-31-16-12 Haywire]$ netstat -nt | grep 8000
tcp        0      0 172.31.16.12:8000           172.31.31.0:36014           CLOSE_WAIT
tcp        0      0 172.31.16.12:8000           172.31.31.0:35808           CLOSE_WAIT
tcp        0      0 172.31.16.12:8000           172.31.31.0:35884           CLOSE_WAIT
tcp        0      0 172.31.16.12:8000           172.31.31.0:36182           CLOSE_WAIT
tcp        0      0 172.31.16.12:8000           172.31.31.0:36267           CLOSE_WAIT
tcp        0      0 172.31.16.12:8000           172.31.31.0:36211           CLOSE_WAIT
tcp        0      0 172.31.16.12:8000           172.31.31.0:35818           CLOSE_WAIT
tcp        0      0 172.31.16.12:8000           172.31.31.0:35940           CLOSE_WAIT
tcp        0      0 172.31.16.12:8000           172.31.31.0:35831           CLOSE_WAIT
tcp        0      0 172.31.16.12:8000           172.31.31.0:35912           CLOSE_WAIT
tcp        0      0 172.31.16.12:8000           172.31.31.0:35849           CLOSE_WAIT
tcp        0      0 172.31.16.12:8000           172.31.31.0:36081           CLOSE_WAIT
tcp        0      0 172.31.16.12:8000           172.31.31.0:35791           CLOSE_WAIT
tcp        0      0 172.31.16.12:8000           172.31.31.0:35963           CLOSE_WAIT
tcp        0      0 172.31.16.12:8000           172.31.31.0:36193           CLOSE_WAIT
tcp        0      0 172.31.16.12:8000           172.31.31.0:36075           CLOSE_WAIT
tcp        0      0 172.31.16.12:8000           172.31.31.0:35872           CLOSE_WAIT
tcp        0      0 172.31.16.12:8000           172.31.31.0:35894           CLOSE_WAIT
tcp        0      0 172.31.16.12:8000           172.31.31.0:35793           CLOSE_WAIT
```

I fixed it by applying the following changes:
```diff
--- a/src/haywire/http_request.c
+++ b/src/haywire/http_request.c
@@ -68,6 +68,7 @@ http_request* create_http_request(http_connection* connection)
     request->body = malloc(sizeof(hw_string));
     request->body->value = NULL;
     request->body->length = 0;
+    request->keep_alive = 0;
     connection->current_header_key_length = 0;
     connection->current_header_value_length = 0;
     INCREMENT_STAT(stat_requests_created_total);
diff --git a/src/haywire/http_server.c b/src/haywire/http_server.c
index 0f64ff0..8ac3a0e 100644
--- a/src/haywire/http_server.c
+++ b/src/haywire/http_server.c
@@ -128,8 +128,12 @@ void free_http_connection(http_connection* connection)
     {
         free_http_request(connection->request);
     }
+
+    for (int i=0; i<connection->response_buffers_count; i++)
+    {
+            free(connection->response_buffers[i].base);
+    }

-    free(connection->response_buffers);
     free(connection);
     INCREMENT_STAT(stat_connections_destroyed_total);
 }
@@ -281,7 +285,7 @@ void http_stream_on_read_http_parser(uv_stream_t* tcp, ssize_t nread, const uv_b
             /* uv_close((uv_handle_t*) &client->handle, http_stream_on_close); */
         }
     }
-    else
+    else if (nread == UV_EOF)
     {
         uv_close((uv_handle_t*) &connection->stream, http_stream_on_close);
     }
@@ -357,6 +361,8 @@ void http_stream_on_read_pico(uv_stream_t* tcp, ssize_t nread, const uv_buf_t* b
             connection->request_buffer_length = 0;
             connection->prevbuflen = 0;
         }
+    } else {
+        uv_close((uv_handle_t*) &connection->stream, http_stream_on_close);
     }
```

In addition to closing the stream, I initialized the `keep_alive` flag after having seen with valgrind that `program.c` was making a conditional based on an initialized value. That meant that `"Connection: keep-alive"` was being written on every response even if the client didn't send the header, resulting in a longer write time per request. 

However, in practice the connection is being kept alive anyway, due to https://github.com/kellabyte/Haywire/blob/picohttpparser/src/haywire/http_server.c#L331. I suppose that's because pico parser doesn't have a http_should_keep_alive function yet. Something to fix before merging :blush: 

Here are some results:

### Without closing the connection on EOF

```
[ec2-user@ip-172-31-31-0 ~]$ ~/wrk/wrk --script pipelined_get.lua --latency -d 10s -t 40 -c 512 http://172.31.16.12:8000 -- 64
Running 10s test @ http://172.31.16.12:8000
  40 threads and 512 connections
  Thread Stats   Avg      Stdev     Max   +/- Stdev
    Latency   113.90ms  207.47ms   1.82s    85.68%
    Req/Sec    43.26k    23.27k  221.44k    92.82%
  Latency Distribution
     50%    9.64ms
     75%  171.65ms
     90%  406.09ms
     99%    0.00us
  17215808 requests in 10.10s, 2.53GB read
  Socket errors: connect 0, read 0, write 0, timeout 5
Requests/sec: 1704452.92
Transfer/sec:    256.81MB
```

### Closing the connection on the server on EOF (keep_alive = 0)

This is with `keep_alive = 0` in `create_http_request` (https://github.com/kellabyte/Haywire/blob/picohttpparser/src/haywire/http_request.c#L62).

```
[ec2-user@ip-172-31-31-0 ~]$ ~/wrk/wrk --script pipelined_get.lua --latency -d 10s -t 40 -c 512 http://172.31.16.12:8000 -- 64
Running 10s test @ http://172.31.16.12:8000
  40 threads and 512 connections
  Thread Stats   Avg      Stdev     Max   +/- Stdev
    Latency     9.19ms   14.59ms 662.40ms   93.51%
    Req/Sec    51.66k    23.89k  239.36k    94.48%
  Latency Distribution
     50%    9.01ms
     75%   13.26ms
     90%   15.96ms
     99%    0.00us
  20493824 requests in 10.10s, 2.56GB read
Requests/sec: 2029286.58
Transfer/sec:    259.33MB
```

### Closing the connection on the server on EOF (keep_alive = 1)

This is with `keep_alive = 1` in `create_http_request` (https://github.com/kellabyte/Haywire/blob/picohttpparser/src/haywire/http_request.c#L62), resulting in the Connection header being generated.

```
[ec2-user@ip-172-31-31-0 ~]$ ~/wrk/wrk --script pipelined_get.lua --latency -d 10s -t 40 -c 512 http://172.31.16.12:8000 -- 64
Running 10s test @ http://172.31.16.12:8000
  40 threads and 512 connections
  Thread Stats   Avg      Stdev     Max   +/- Stdev
    Latency   112.43ms  201.63ms   1.91s    85.01%
    Req/Sec    43.16k    23.07k  216.32k    93.71%
  Latency Distribution
     50%    9.81ms
     75%  176.50ms
     90%  409.75ms
     99%    0.00us
  17222408 requests in 10.10s, 2.53GB read
  Socket errors: connect 0, read 0, write 0, timeout 4
Requests/sec: 1705117.50
Transfer/sec:    256.93MB
```

This pretty much equivalent to the version without my changes :blush: